Doc: Add tutorial about using zephyr as uos (#2721)
projectacrn/acrn-hypervisor: doc/tutorials/using_zephyr_as_uos.rst

.. _using_zephyr_as_uos:

Using Zephyr as User OS
#######################

> The tutorial describes the steps to run zephyr as User OS on ACRN hypervisor. We are using

Reviewer: ```suggestion
This tutorial describes how to run Zephyr as the User OS on the ACRN hypervisor. We are using
```

Author: Will fix it.

> Kaby Lake-based NUC model NUC7i5DNHE and you can try other platform ACRN supported as well.

Reviewer: ```suggestion
Kaby Lake-based NUC (model NUC7i5DNHE) in this tutorial.
Other :ref:`ACRN supported platforms <hardware>` should work as well.
```

Author: Will fix it.

Introduction to Zephyr
**********************

> The Zephyr is a scalable real-time operating-system (RTOS) supporting multiple hardware architecture,

Reviewer: ```suggestion
The Zephyr RTOS is a scalable real-time operating-system supporting multiple hardware architectures,
```

Author: Will fix it.

optimized for resource constrained devices, and built with safety and security in mind.

Steps for Using Zephyr as User OS
*********************************

#. Build Zephyr

>    First of all, please follow the `Getting Started Guide

Reviewer: ```suggestion
   Follow the `Zephyr Getting Started Guide
```

Author: Will fix it.

   <https://docs.zephyrproject.org/latest/getting_started/>`_ to

>    setup Zephyr development environment. Then follow the `Board-Specfic Support

Reviewer: ```suggestion
   setup the Zephyr development environment. Then follow the `Board-Specfic Support
```

Author: Will fix it. Thanks.

>    <https://docs.zephyrproject.org/latest/boards/index.html>`_ to build the acrn supported zephyr image ``zephyr.elf``.

Reviewer: ```suggestion
   <https://docs.zephyrproject.org/latest/boards/index.html>`_
   to build the Zephyr image ``zephyr.elf``.
```

Author: Will fix it. Thanks.

>    .. note::

Reviewer: This note should not be included in the documentation.  We can't publish this doc until the Zephyr PR is merged.

Author: Got it.

      When the `Getting Started Guide` for building acrn supported zephyr image got merged to `Zephyr Tree
      <https://github.com/zephyrproject-rtos/zephyr>`_, this section will be update to the exact reference.
  
#. Build grub2 boot loader image

>    We can build grub2 bootloader for zephyr using ``boards/x86/common/scripts/build_grub.sh``

Reviewer: The correct spelling is ``Zephyr`` with an uppercase Z

Author: Will fix it.

   which locate in `Zephyr Sourcecode <https://github.com/zephyrproject-rtos/zephyr>`_.

>    .. code-block:: none

Reviewer: The extra space is correct (this code-block should be indented to be part of the ordered list).

>    .. code-block:: none

Reviewer 2: > The extra space is correct (this code-block should be indented to be part of the ordered list).

Do we really need to `<space>` characters at the beginning of lines 33 and 43? There is no text at all on those (I was not referring to the indentation of the `code-block` directive itself.

      $ ./boards/x86/common/scripts/build_grub.sh x86_64

>    Find the binary at boards/x86/common/scripts/grub/bin/grub_x86_64.efi.

Author: Will fix it.

#. Preparing the boot device

   .. code-block:: none

>       $ dd if=/dev/zero of=zephyr.img bs=1M count=1024

Reviewer 2: This creates a 1GB image, and we don't even need 1.5M with all the files. Let's create a smaller one:
```suggestion
      $ dd if=/dev/zero of=zephyr.img bs=1M count=35
```

**Note:** I am using 35MB on purpose here as this is the smallest partition size allowed by `FAT32`.

Author: Yes, will do.

      $ mkfs.vfat -F 32 zephyr.img

>       $ losetup -f -P --show zephyr.img

Author: Thanks, will do.

>       $ mount /dev/loop0 /mnt

Author: Thanks, will fix it.

>       $ mount /dev/loop0 /mnt

Reviewer 2: I forgot that it would be better to account for the fact that the loop device assigned may be different than `lopp0` here, can you change these couple of lines to:
```
$ LOOP_DEV=`sudo losetup -f -P --show zephyr.img`
$ sudo mount $LOOP_DEV /mnt
```

A single-line command is also possible:
```
$ sudo mount `sudo losetup -f -P --show zephyr.img` /mnt
```

   Create the following directories.

>    :file:`/mnt/efi`

Author: Suggestion accepted.

   :file:`/mnt/efi/boot`

   :file:`/mnt/kernel/`

   Copy ``zephyr.elf`` to ``/mnt/kernel/`` and ``grub_x86_64.efi`` to ``/mnt/efi/boot/``.

   Create ``/mnt/efi/boot/grub.cfg`` containing the following:

   .. code-block:: console

     set default=0
     set timeout=10

     menuentry "Zephyr Kernel" {
         multiboot /kernel/zephyr.elf
     }

   .. note::
      ``zephyr.elf`` locates in the ``build/zephyr/``. For example, we can find ``zephyr.elf`` in
      ``samples/hello_world/build/zephyr/zephyr.elf`` if we build a ``Hello World App``.

#. Follow `Getting started guide for Intel NUC <https://projectacrn.github.io/latest/getting-started/apl-nuc.html>`_ to
   boot "The ACRN Service OS"

#. Boot Zephyr as User OS

>    Make one new directory ``zephyr`` in SOS. And then copy zephyr.img, ``/usr/share/acrn/samples/nuc/launch_zephyr.sh``

Reviewer 2: I don't think we install the `launch_zephyr.sh` script in the Clear Linux packaging, @miguelinux can you confirm this and fix it if true? Thanks!

>    and ``/usr/share/acrn/samples/bios/`` to ``zephyr`` directory. We can find ``OVMF.fd`` and ``launch_zephyr.sh`` in

Reviewer 2: ```suggestion
   and ``/usr/share/acrn/bios/`` to ``zephyr`` directory. We can find ``OVMF.fd`` and ``launch_zephyr.sh`` in
```

Author: Will fix it. Thanks.

   `ACRN Repo <https://github.com/projectacrn/acrn-hypervisor>`_ also.

   Run the ``launch_zephyr.sh`` script to launch Zephyr as UOS.

   .. code-block:: none

>       $ ./launch_zephyr.sh

Reviewer 2: You need to be sudo here:
```suggestion
      $ sudo ./launch_zephyr.sh
```

Author: Will fix it.

   Load Zephyr using ``grub_x86_64.efi`` in UEFI shell:

   .. code-block:: console

      $ shell> FS0:
      $ FS0:\> efi\boot\grub_x86_64.efi

   Then zephyr will boot up automatically. You will see the prompt as following:

   .. code-block:: console

>       uart:~$

Reviewer 2: I have not been able to get to that prompt. I tried to build Zephyr for the `up_squared`, `qemux86` and `qemux86_64` boards but none worked for me. Which board did you use to build the `samples/hello_world` Zephyr app?

Author: @gvancuts, there is patch which should be applied on the top of zephyr tree.
And here is an closed PR you can take as an reference. https://github.com/zephyrproject-rtos/zephyr/pull/14271.

Charles will restructuring the patch and PR again.  After the PR merged, I will update the section
"Build Zephyr" accordingly.

Reviewer 2: Thanks!

.. _using_zephyr_as_uos:

Using Zephyr as User OS
#######################

The tutorial describes the steps to run zephyr as User OS on ACRN hypervisor. We are using

> Kaby Lake-based NUC model NUC7i5DNH and you can try other platform ACRN supported as well.

Author: Thanks for your comment. Will address it next version.

Introduction to Zephyr
**********************

The Zephyr is a scalable real-time operating-system (RTOS) supporting multiple hardware architecture,
optimized for resource constrained devices, and built with safety and security in mind.

Steps for Using Zephyr as User OS
*********************************

#. Build Zephyr and grub2 boot loader image

   First of all, please follow the `Development Environment Setup on Linux

>    <https://docs.zephyrproject.org/latest/getting_started/installation_linux.html>`_ to

Reviewer 2: I think it would be better to point at the [Getting Started Guide](https://docs.zephyrproject.org/latest/getting_started/) as opposed to just that one page. There are instructions below that will require the user to have gone through most of these pages (e.g. `cd $ZEPHYR_BASE`).

Author: Sure. Will update it in next version.

   setup Zephyr development environment. Then follow the `Board-Specfic Support

>    <https://docs.zephyrproject.org/latest/boards/index.html>`_ to build the acrn supported zephyr image ``zephyr.elf``.

Reviewer 2: There is no such reference to `zephyr.elf` on this page (or even some sort of an ACRN BSP page). Can you add a more accurate pointer to where one can find info about the next steps?

@dbkinder , I'm actually thinking that having an `ACRN` bsp page on Zephyr would be good, what do you think?

Author: @gvancuts, there are patches in zephyr side containing one doc about how to build zephyr.elf which can run on top of acrn. I will update the reference with that doc once those patches got merged in zephyr upstreaming tree.

Reviewer 2: Great! Do you have a link to the PR yet?

   Build grub2 boot loader image

   .. code-block:: none

      $ git clone https://github.com/zephyrproject-rtos/zephyr.git

>       $ cd $ZEPHYR_BASE

Reviewer 2: This does not make much sense in this sequence, `ZEPHYR_BASE` is not defined just after the `git clone`. I think the best is to refer to the Getting Started Guide and ask people to go through the various steps, such as installing the host tools, pick a Zephyr SDK, configure their `~/.zephyrrc`, ecc... and only mention the script below in this doc.

Author: Ok. Will do.

>       $ ./boards/x86/common/scripts/build_grub.sh x86_64

Reviewer 2: I'm getting an error on my system (`Clear Linux version: 27670`) when running this step:
```
In file included from grub-core/partmap/gpt.c:26:
./include/grub/gpt_partition.h:79:1: error: alignment 1 of ‘struct grub_gpt_partentry’ is less than 8 [-Werror=packed-not-aligned]
 } GRUB_PACKED;
 ^
cc1: all warnings being treated as errors
make[2]: *** [Makefile:6490: grub-core/partmap/libgrubkern_a-gpt.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

I have tried with both Zephyr SDK 0.9.5 and 0.10.0, will try shortly upgrading my Clear Linux host to see if it makes any difference.

Have you seen this error before?

Author: I met the same error on clearlinux. But it works on ubuntu-16.04.
How about add one note here to remind user to use ubuntu-1604?

Reviewer 2: Good to know it wasn't just me :-)

I have filed a bug in Zephyr to track the issue on Clear Linux: https://github.com/zephyrproject-rtos/zephyr/issues/14289

Because of that issue, I used Fedora 29 to build it but had problems with it too. It was easy enough to resolve them though, a PR is pending with the fixes here: https://github.com/zephyrproject-rtos/zephyr/pull/14281

I have not tested other versions but I would simply add a note for now indicating there is a known issue with Clear Linux (and a link to it).

Reviewer 2: https://github.com/zephyrproject-rtos/zephyr/pull/14468 landed recently in Zephyr. Fedora 29 and Clear Linux can now build the GRUB2 firmware image correctly, so no need to insert any note anymore :-)

   Find the binary at boards/x86/common/scripts/grub/bin/grub_x86_64.efi.

#. Preparing the boot device

   .. code-block:: none

      $ dd if=/dev/zero of=zephyr.img bs=1M count=1024
      $ mkfs.vfat -F 32 zephyr.img
      $ losetup -f -P --show zephyr.img
      $ mount /dev/loop0 /mnt

   Create the following directories.

   :file:`/mnt/efi`

   :file:`/mnt/efi/boot`

   :file:`/mnt/kernel/`

>    Copy ``zephyr.elf`` to ``/mnt/kernel/`` and ``grub_x86_64.efi`` to ``/mnt/efi/boot/``.

Reviewer 2: Where are you getting the `zephyr.elf` file from? I'm assuming you are building a sample app or something?

Author: Yes.
If we build samples/hello_world, zephyr.elf will locate at samples/hello_world/build/zephyr/zephyr.elf.

If we build zephyr with shell, zephyr.elf will locate at samples/subsys/shell/shell_module/build/zephyr/zephyr.elf

   Create ``/mnt/efi/boot/grub.cfg`` containing the following:

   .. code-block:: console

     set default=0
     set timeout=10

     menuentry "Zephyr Kernel" {
         multiboot /kernel/zephyr.elf
     }

#. Follow `Getting started guide for Intel NUC <https://projectacrn.github.io/latest/getting-started/apl-nuc.html>`_ to
   boot "The ACRN Service OS"

#. Boot Zephyr as User OS

>    Copy the launch script ``devicemodel/samples/nuc/launch_zephyr.sh``, ``devicemodel/bios/OVMF.fd`` and ``zephyr.img``

Reviewer 2: This kind of implies the user did a `git clone ...` for the ACRN repo but that's not in this guide. Assuming the user has followed the ACRN GSG steps he should actually have the `launch_zephyr.sh` scripts on his filesystem already, same for the `OVMF.fd` file so we should simply instruct the user to pick these up from `/usr/share/acrn/samples/nuc/` and `/usr/share/acrn/samples/bios/` (going from memory for the exact paths, so to be checked).

Author: Yes. It is a better way. I will update it.

   to ``SOS:$HOME/``.

   .. code-block:: none

      $ ./launch_zephyr.sh

   Load Zephyr using ``grub_x86_64.efi`` in UEFI shell:

   .. code-block:: console

>       $ shell> FS0:

Reviewer 2: I'll need to resolve my issue building grub first but I am surprised by this step, isn't that done automatically at boot time? I mean, doesn't `OVMF` find the Grub bootloader and load it?

Author: Here we should load bootloader menually. Do you mean that it is better to let OVMF load
the bootloader automatically?

Reviewer 2: Yes, I don't think it helps much to have the user manually go and load grub. You _just_ need to do this to have it start loaded automatically by OVMF:
```
sudo cp boards/x86/common/scripts/grub/bin/grub_x86_64.efi /mnt/efi/boot/bootx64.efi
```
And with that, when you start the VM, it will automatically go to the Grub menu where you can choose to start Zephyr.
